### DISCUSSION TITLE
Importing seeders for ES Modules

### DISCUSSION THREAD

**Discussion by ceelogre** [sequelize/umzug#360] [Q&A] [answered]
I have `type:module` in my project so, when I generate seeders, they are ES Modules.
When I try to seed, it seems `umzug` is unable to load the ES Module file.
```javascript  
Must use import to load ES Module: /../20201013130912-user.js.  
require() of ES modules is not supported. 
Instead rename 20201013130912-user.js to end in .cjs, change the requiring code to use import(), or remove "type": "module"
 from package.json
 ```  

Obviously, I can do the renaming but I was wondering if there's a way to do this automatically or another fix, cause I intend to push this to a build tool where I may not be able to rename in the middle of a build.

**Note**: Downgrading my project to Common JS is not an option for now.

**Comment by mmkal** [comment]
@ceelogre the readme has some instructions for doing this with typescript modules. Basically at the entry point for your migration script, you call `require('ts-node/register')` and from then on it's possible to `require` typescript modules, which is what umzug does internally. If you're not using typescript, there are a couple of options:
- you could install typescript and add a `tsconfig.json` to your project with `allowJs` set to true, then ts-node will be able to handle your (js) ES modules. Or, you might have some luck with `require('ts-node/register/transpile-only')`
- I haven't tried this, but if you're using babel it might be easier to fit into your existing setup, use [babel register](https://babeljs.io/docs/en/babel-register) in a similar way.

P.S. I moved this to discussions as I don't think there's a bug here, but the end result could be something added to the docs.

**Reply by ceelogre** [reply]
I'm not using any TypeScript and I've moved away from babel since most of it is now supported natively by Node. I'll keep looking for another solution. 

Good idea on moving this to Discussion.

**Reply by mmkal** [reply]
Got it. Have you tried using `import(...)` in a custom resolver?

**Reply by ceelogre** [reply]
Not quite sure how to go about that. Guidance would be appreciated :)

**Reply by mmkal** [reply]
Sure - could you share how you're creating an umzug instance? Also which umzug version you're using. There are breaking changes between v2 and v3 (beta).

**Reply by ceelogre** [reply]
I'm not calling the instance directly, it's called when I run `npx sequelize-cli seed:generate --name user` this generate a `.js` file which fails to import when I run `npx sequelize-cli db:seed:all`

**Reply by mmkal** [reply]
Ah, I'm not familiar with sequelize-cli. @papb do you know if it's possible to use custom resolvers with it?

**Reply by mmkal** [reply]
@ceelogre curious if you ever made progress on this? I'd probably lean towards suggesting just not using sequelize-cli if it doesn't support ES modules. There may be an umzug-cli which is database-agnostic (i.e. no dependency on or coupling to sequelize). Would that be of interest?

**Reply by ceelogre** [reply]
Hi @mmkal, I did make progress but it was more of a hack. Basically, I have to generate the model using sequelize-cli, edit it to add my own attributes, then rename it to end in `.cjs` since it's commonjs file. The process is not really efficient, I'd love to hear about the umzug-cli that is database-agnostic.

**Reply by mmkal** [reply]
@ceelogre see https://github.com/sequelize/umzug/pull/389. I published a prerelease CLI package:

```
npm i umzug@beta-cli
```

It'd be great to get your feedback/to find out if this works with ES modules.

**Reply by papb** [reply]
> 
> 
> Ah, I'm not familiar with sequelize-cli. @papb do you know if it's possible to use custom resolvers with it?

Currently not possible... I plan on overhauling `sequelize-cli` once Umzug v3 is out, even more now that umzug is providing an amazing CLI.

**Comment by ceelogre** [comment]
Good to hear. I'll check it out tomorrow.

Télécharger Outlook pour Android<https://aka.ms/ghei36>

________________________________

**Comment by mmkal** [comment, marked as answer]
@ceelogre CLI is now published on the beta tag (`npm install umzug@beta`) and there's an example of how you can use it with es modules here: https://github.com/sequelize/umzug/tree/master/examples/2.es-modules

Note that there may be slight changes to the API before the next major release. For this question you may also be interested in looking at the [seeders example](https://github.com/sequelize/umzug/tree/master/examples/4.sequelize-seeders).

**Reply by ceelogre** [reply]
Thanks for taking it into consideration.
